internal/transports/p2p/server.go

```diff
@@ -29,99 +43,271 @@ func NewServer(
 	log *zerolog.Logger,
 ) *server {
 	serverLogger := log.With().Str("service", "p2p-experimental").Logger()
+	ctx, ctxCancel := context.WithCancel(context.Background())
 	server := &server{
 		config:         config,
 		chainParams:    chainParams,
 		headersService: headersService,
 		chainService:   chainService,
 		log:            &serverLogger,
-		peers:          make([]*peer.Peer, 0),
+		outboundPeers:  peer.NewPeersCollection(config.MaxOutboundConnections),
+		inboundPeers:   peer.NewPeersCollection(config.MaxInboundConnections),
+		addresses:      network.NewAdressBook(config.BanDuration, config.AcceptLocalPeers),
+		ctx:            ctx,
+		ctxCancel:      ctxCancel,
 	}
+
 	return server
 }
 
+// Start starts the P2P server by connecting to outbound peers and listening for inbound connections.
 func (s *server) Start() error {
-	err := s.seedAndConnect()
+	err := s.connectOutboundPeers()
 	if err != nil {
 		return err
 	}
 
-	return s.listenAndConnect()
+	err = s.listenInboundPeers()
+	if err != nil {
+		s.log.Error().Msgf("error during server start. Shutdown p2p server. Reason: %v", err)
+		s.Shutdown()
+
+		return err
+	}
```

Comment:
1. Methods `connectOutboundPeers` and `listenInboundPeers` could be goroutines, but this can also be a next task to do for this package. WDYT?
2. We only shutdown when `listenInboundPeers` fails, but we don't shutdown when `connectOutboundPeers` fails? I think `server.Shutdown()` should happen in `main.go` when `server.Start()` returns an error. So I would remove the shutdown from here completely.